### PR TITLE
Vb 3476 link to api endpoint using sub to bring back related prisoner info

### PR DESCRIPTION
**Pull request by dhirajshettymoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      description = "Booker's unique reference.",
+      example = "A12345DC",
+    )
+    @NotBlank
+    bookerReference: String,
+  ): List<AssociatedPrisonerDto> {
+    return prisonerDetailsService.getAssociatedPrisoners(bookerReference)
+  }
+
+  @PreAuthorize("hasRole('PUBLIC_VISITS_BOOKING_SERVICE')")
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt:68]: Is this the role we are going to use I thought we had change it?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+const val BOOKER_LINKED_PRISONERS: String = "$PUBLIC_BOOKER_CONTROLLER_PATH/prisoners"
+const val BOOKER_LINKED_PRISONER_VISITORS: String = "$BOOKER_LINKED_PRISONERS/{prisonerId}/visitors"
+
+@RestController
+class OrchestrationPublicBookerController(
+  val prisonerDetailsService: PrisonerDetailsService,
+  val visitorDetailsService: VisitorDetailsService,
+) {
+  @PreAuthorize("hasRole('PUBLIC_VISITS_BOOKING_SERVICE')")
+  @GetMapping(BOOKER_LINKED_PRISONERS)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt:30]: Is this the role we are going to use I thought we had change it?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/entity/BookerPrisoner.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  @GeneratedValue(strategy = GenerationType.IDENTITY)
+  @Column(name = "ID")
+  val id: Long = 0,
+
+  @Column(name = "booker_id", nullable = false)
+  val bookerId: Long,
+
+  @ManyToOne
+  @JoinColumn(name = "booker_id", updatable = false, insertable = false)
+  val booker: Booker,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/entity/BookerPrisoner.kt:31]: Were is the other part of this relationship in the Booker
`val prisoners: MutableList<BookerPrisoner> = mutableListOf(),`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/dto/orchestration/PrisonerBasicInfoDto.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package uk.gov.justice.digital.hmpps.oneloginuserregistry.dto.orchestration
+
+import io.swagger.v3.oas.annotations.media.Schema
+import java.time.LocalDate
+
+data class PrisonerBasicInfoDto(
+  @Schema(required = true, description = "Prisoner Number", example = "A1234AA")
+  val prisonerNumber: String,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/dto/orchestration/PrisonerBasicInfoDto.kt:8]: nice to have required at end, like we have done in other projects

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/repository/BookerRepository.kt [hunk 5]
```diff
@@ -9,5 +9,7 @@ interface BookerRepository : JpaRepository<Booker, Long> {
 
   fun findByOneLoginSub(authReference: String): Booker?
 
+  fun findByReference(reference: String): Booker?
+
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/repository/BookerRepository.kt:13]: should never be null as you have reference

Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/repository/BookerRepository.kt:13]: Booker?

**Review by AledWEvans — CHANGES_REQUESTED** [review]
Some minor comments

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt:68]: oh yes, will update the role

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/dto/orchestration/PrisonerBasicInfoDto.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+  @Schema(required = true, description = "Last name", example = "Larsen")
+  val lastName: String,
+
+  @Schema(required = true, description = "Date of Birth", example = "1975-04-02")
+  val dateOfBirth: LocalDate?,
+
+  @Schema(description = "Prison ID", example = "MDI")
+  val prisonId: String?,
+)
```
[the remark this answers is not in the document]
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/dto/orchestration/PrisonerBasicInfoDto.kt:21]: I have added them for now but not populating them. Maybe I will remove them for now and add them back when these might be needed.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  private val bookerRepository: BookerRepository,
+  private val prisonerRepository: BookerPrisonerRepository,
+  private val orchestrationServiceClient: OrchestrationServiceClient,
+) {
+  companion object {
+    private const val NOT_KNOWN = "NOT_KNOWN"
+  }
+
+  fun getAssociatedPrisoners(reference: String): List<AssociatedPrisonerDto> {
+    val bookerByReference = bookerRepository.findByReference(reference)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt:22]: throw not found exception

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+          associatedPrisoners.add(AssociatedPrisonerDto(prisonerDetails[it.prisonNumber] ?: getBlankPrisonerBasicInfo(it.prisonNumber), it))
+        }
+      }
+
+      associatedPrisoners
+    } ?: emptyList()
+  }
+
+  fun getAssociatedPrisoner(reference: String, prisonerId: String): BookerPrisoner? {
+    val bookerByReference = bookerRepository.findByReference(reference)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt:40]: throw not found exception

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt]: updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/controller/BookerDetailsController.kt:30]: updated to BOOKER_AUTHORISATION

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/entity/BookerPrisoner.kt (hunk 3 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/entity/BookerPrisoner.kt:31]: thanks, added to BOOKER.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/dto/orchestration/PrisonerBasicInfoDto.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/dto/orchestration/PrisonerBasicInfoDto.kt:7]: Updated

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/repository/BookerRepository.kt (hunk 5 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/model/repository/BookerRepository.kt:13]: as discussed, throwing BookerNotFoundException if reference is not valid. Also added tests.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt (hunk 7 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt]: throwing BookerNotFoundException

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt (hunk 8 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt]: throwing BookerNotFoundException

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  private val prisonerRepository: BookerPrisonerRepository,
+  private val orchestrationServiceClient: OrchestrationServiceClient,
+) {
+  companion object {
+    private const val NOT_KNOWN = "NOT_KNOWN"
+  }
+
+  fun getAssociatedPrisoners(reference: String): List<AssociatedPrisonerDto> {
+    val bookerByReference = bookerRepository.findByReference(reference) ?: throw BookerNotFoundException("Booker for reference : $reference not found")
+    val associatedPrisoners = mutableListOf<AssociatedPrisonerDto>()
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oneloginuserregistry/service/PrisonerDetailsService.kt:24]: This and the other call could be extracted into a method

```
fun getBooker(reference: String) : Booker {
           return bookerRepository.findByReference(reference) ?: throw BookerNotFoundException("Booker for reference : $reference not found")
}
```

**Review by AledWEvans — APPROVED** [review]
LGTM

**Review by AledWEvans — APPROVED** [review]
LGTM

**Pull request merged**